cabal-install/Distribution/Client/Dependency/TopDown.hs

```diff
@@ -127,7 +127,8 @@ explore pref (ChoiceNode _ choices)  =
       where
         isInstalled (SourceOnly _) = False
         isInstalled _              = True
-        isPreferred p = packageVersion p `withinRange` preferredVersions
+        isPreferred p = packageVersion p `withinRange`
+                        combinePrefRanges preferredVersions
         (PackagePreferences preferredVersions packageInstalledPreference)
```

Comment:
The comment in `resolveWithoutDependencies` also applies here, and on the next change in this file.  I think that it would be better to either keep the old behavior or use the same comparison as in the modular solver, as long as the change isn't much larger.